Automattic/newspack-ads
# feat: global placements

![image](https://user-images.githubusercontent.com/820752/141354073-599e0414-8e21-44c2-a933-4611bde9e86b.png)

This PR ports and refactors the Ad Placement system from the core plugin.

As proposed here, placements are now configurable through the `newspack_ads_placements` filter and should be set with the following:

- name: The name of the placement.
- description: A description of the placement.
- default_enabled: Whether this placement should be enabled by default.
- default_ad_unit: A default ad unit name to be used for this placement.
- hook_name: The name of the WordPress action hook to inject an ad unit into.
- hooks: An array of action hooks to inject an ad unit into.
  - name: The friendly name of the hook.
  - hook_name: The name of the WordPress action hook to inject the ad unit into.

It implements by default the existing placements with backward compatibility and deprecates the selection of service providers since it just connects with GAM and will continue to do so.

## Widget area placements

The `Newspack_Ads_Sidebar_Placements` class makes use of the API provided by `Newspack_Ads_Placements` to configure placements for the widget areas. By default, all sidebars are hooked as placement areas, implementing 2 hooks for inserting ads `before` and `after` the widgets.

It implements two filters targeted for customization according to the theme:

- `newspack_ads_disallowed_sidebar_placements`:  Sidebars to not create ad placement. Initially populated with Newspack sidebars that are already placed through the core global placement.
- `newspack_ads_single_unit_sidebar_placements`: Sidebars that should have a single ad unit instead of "before" and "after" hooks. Initially populated with the Newspack `Above Copyright` sidebar.

On the image below, the **Sidebar** with 2 `hooks` for "before" and "after" and the **Above Copyright** using only the main ``hook_name``.

Note: Sidebars without widgets will not have ad units inserted.

![image](https://user-images.githubusercontent.com/820752/141356243-8fc80351-c6f1-4156-997a-0941c4abe079.png)

## SCAIP

SCAIP uses widget areas for ad placement, but they are also skipped here. These placements will be handled in a different PR because it requires a specific logic for backward compatibility and avoid duplication of the Ad Unit widget with the placement system proposed here.

## Testing this PR

### Backward compatibility

1. With this plugin and https://github.com/Automattic/newspack-plugin at `master` branch, set ad units for all available placements.
2. Switch to this branch and https://github.com/Automattic/newspack-plugin/pull/1225, and run `npm run build` for `newspack-plugin`
3. Visit the `Advertising > Placements` wizard and confirm all the ad units remained selected and working
4. Select a different ad unit for the placements, visit your website and make sure it is stored and updated.

### Testing widget areas

1. On the wizard, activate "Sidebar"
2. Set ad units for "Before widgets" and "After widgets"
3. Confirm they are visible on the sidebar
4. Activate "Above Copyright" and make sure you don't have widgets on this sidebar
5. Set an ad unit and confirm it is **not** visible
6. Add a widget to the "Above Copyright" sidebar, refresh your page and confirm the ad unit is now visible
